pkg/storegateway/bucket_index_reader.go

```diff
@@ -62,8 +79,10 @@ func newBucketIndexReader(block *bucketBlock) *bucketIndexReader {
 // single label name=value.
 func (r *bucketIndexReader) ExpandedPostings(ctx context.Context, ms []*labels.Matcher, stats *safeQueryStats) (returnRefs []storage.SeriesRef, returnErr error) {
 	var (
-		loaded bool
-		cached bool
+		loaded           bool
+		cached           bool
+		promise          expandedPostingsPromise
+		deferredMatchers []*labels.Matcher
```

Comment:
This is a late and opinionated feedback, but I'd consider calling this `pendingMatchers`: I think it would be easier to understand for newcomers to this codebase as these are matchers pending to be applied.

Reply:
Will continue in the previous thread https://github.com/grafana/mimir/pull/4667#discussion_r1158765186